spyder_kernels/utils/iofuncs.py

```diff
@@ -370,6 +369,25 @@ def save_dictionary(data, filename):
     return error_message
 
 
+def is_within_directory(directory, target):
+    """Check if a file is within a directory."""
+    abs_directory = os.path.abspath(directory)
+    abs_target = os.path.abspath(target)
+    prefix = os.path.commonprefix([abs_directory, abs_target])
+    return prefix == abs_directory
+
+
+def safe_extract(tar, path=".", members=None, numeric_owner=False):
+    """Safely extract a tar file."""
+    for member in tar.getmembers():
+        member_path = os.path.join(path, member.name)
+        if not is_within_directory(path, member_path):
+            raise Exception(
+                r"Attempted path traversal in tar file {}".format(tar.name)
```

Comment:
```suggestion
                f"Attempted path traversal in tar file {tar.name!r}"
```

* Not sure why this is a raw string, as no escapes are used
* May as well use an f-string instead of `format`?
* Using the `r` (repr) format code will quote the file name

Reply:
> Not sure why this is a raw string, as no escapes are used

Sorry, I thought the `!r` meant raw string.

> May as well use an f-string instead of format?

The 2.x branch still support Python 2, so that's not possible.

> Using the r (repr) format code will quote the file name

Good to know! That's a change I can make.

Reply:
> Sorry, I thought the !r you suggested in other PR meant raw string.

`!` indicates format codes in the format-string mini language, analogous to `%f`, `%s`, etc. in printf/`%`-style formatting

> The 2.x branch still support Python 2, so that's not possible.

Oh, didn't think of that even though the code had a `PY2` branch; its been a while since I've written Python 2 compatible code, and even if the function is never called on Py2, the unrecognized string prefix will presumably raise a SyntaxError in the tokenizer or the parser before it is compiled to bytecode and the AST won't be valid. Good catch.

> Good to know! That's a change I can make.

What I meant was using the proper format code as I did above, i.e.:

```python
raise Exception("Attempted path traversal in tar file {!r}".format(tar.name))
```

Like the other `format` syntax, it works the same whether you're using f-strings or `str.format()`.